### PR TITLE
crossplane 1.17.1 (new formula)

### PR DESCRIPTION
**Pull request by Moulick**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
https://github.com/crossplane/crossplane/issues/4899

### CODE REVIEW

File: Formula/c/crossplane.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", *std_go_args(ldflags: "-s -w -X=github.com/crossplane/crossplane/internal/version.version=v#{version}", output: bin/"crossplane"), "./cmd/crank"
+  end
+
+  test do
+    assert_match "Client Version: v#{version}", shell_output("#{bin}/crossplane version --client")
+  end
```
Comment (Moulick) [inline Formula/c/crossplane.rb:17]: Adding some proper test will require adding about ~3 yaml files of about 20-30 lines each. Is that required?

File: Formula/c/crossplane.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  homepage "https://github.com/crossplane/crossplane"
+  url "https://github.com/crossplane/crossplane/archive/refs/tags/v1.17.1.tar.gz"
+  sha256 "44e8c94ffa41174f12e25ff60a9c7fd30cfd404b07b2332ec04070a27c2d2e74"
+  license "Apache-2.0"
+  head "https://github.com/crossplane/crossplane.git", branch: "main"
+
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", *std_go_args(ldflags: "-s -w -X=github.com/crossplane/crossplane/internal/version.version=v#{version}", output: bin/"crossplane"), "./cmd/crank"
```
Comment (chenrui333) [inline Formula/c/crossplane.rb:12]: why just build crank, but not crossplane as well?

File: Formula/c/crossplane.rb (hunk 1 above)
Comment (chenrui333) [reply Formula/c/crossplane.rb:17]: maybe download as resource or copy the files from tarball?

Comment (Moulick) [reply Formula/c/crossplane.rb:17]: Else some beta commands like `crossplane beta convert deployment-runtime` can take in a 7 line file like

```yaml
apiVersion: pkg.crossplane.io/v1alpha1
kind: ControllerConfig
metadata:
 name: irsa
spec:
 args:
   - --enable-external-secret-stores
```

and output

```yaml
apiVersion: pkg.crossplane.io/v1beta1
kind: DeploymentRuntimeConfig
metadata:
  name: irsa
spec:
  deploymentTemplate:
    spec:
      selector: {}
      strategy: {}
      template:
        metadata:
        spec:
          containers:
          - args:
            - --enable-external-secret-stores
            name: package-runtime
            resources: {}
```

Is there any formula that  does this sort of file equal assert?

File: Formula/c/crossplane.rb (hunk 2 above)
Comment (Moulick) [reply Formula/c/crossplane.rb:12]: I don't quite get what's the use of crossplane binary for normal users. Following https://docs.crossplane.io/latest/cli/ the shell script referenced installs only the `crank` binary

Comment (phisco) [reply Formula/c/crossplane.rb:12]: Crank is the crossplane cli, crossplane is the actual crossplane binary, so in this context only crank makes sense indeed.

Comment (Moulick) [reply Formula/c/crossplane.rb:12]: I think the `crossplane` binary runs in an Init container to install some crossplane CRDs to a cluster where someone is installing Crossplane via Helm. It does not seems to be intended for running on local machines.

Comment (phisco) [reply Formula/c/crossplane.rb:12]: Both as init container and main container. I confirm it's not meant to be run locally.

Comment (chenrui333) [reply Formula/c/crossplane.rb:12]: sounds good then, just checking 👍

Comment (chenrui333) [reply Formula/c/crossplane.rb:12]: (related, would it make more sense to have the formula as crank, rather than crossplane then?)

Comment (phisco) [reply Formula/c/crossplane.rb:12]: Crank is just an internal name, the cli should be called crossplane, so crossplane is correct 👍

File: Formula/c/crossplane.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        creationTimestamp: "2024-09-29T18:32:47Z"
+        name: irsa
+      spec:
+        deploymentTemplate:
+          spec:
+            selector: {}
+            strategy: {}
+            template:
+              metadata:
+                creationTimestamp: "2024-09-29T18:32:47Z"
```
Comment (Moulick) [inline Formula/c/crossplane.rb:39]: Need to find a way to strip this timestamp/line from the output of the convert command

Comment (Moulick) [reply Formula/c/crossplane.rb]: Nevermind, using `#{time.iso8601}` works!

Comment (Moulick) [reply Formula/c/crossplane.rb]: Nope, don't work, this is just a race condition that fails 😒 . Need to find a way to strip the `creationTimestamp` line from the output `deploymentruntimeconfig.yaml`

File: Formula/c/crossplane.rb (hunk 2 above)
Comment (SMillerDev) [reply Formula/c/crossplane.rb:12]: So `brew install crossplane` will mean that there is no `crossplane` binary, just a `crank` one?

Comment (Moulick) [reply Formula/c/crossplane.rb:12]: `brew install crossplane` will install the `crank` binary as `crossplane`.

File: Formula/c/crossplane.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  homepage "https://github.com/crossplane/crossplane"
+  url "https://github.com/crossplane/crossplane/archive/refs/tags/v1.17.1.tar.gz"
+  sha256 "44e8c94ffa41174f12e25ff60a9c7fd30cfd404b07b2332ec04070a27c2d2e74"
+  license "Apache-2.0"
+  head "https://github.com/crossplane/crossplane.git", branch: "main"
+
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", *std_go_args(ldflags: "-s -w -X=github.com/crossplane/crossplane/internal/version.version=v#{version}", output: bin/"crossplane"), "./cmd/crank"
```
Comment (chenrui333) [inline Formula/c/crossplane.rb:12]: this is not needed
```suggestion
    system "go", "build", *std_go_args(ldflags: "-s -w -X github.com/crossplane/crossplane/internal/version.version=v#{version}"), "./cmd/crank"
```

File: Formula/c/crossplane.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  test do
+    assert_match "Client Version: v#{version}", shell_output("#{bin}/crossplane version --client")
```
Comment (chenrui333) [inline Formula/c/crossplane.rb:16]:
```suggestion
    assert_match "Client Version: v#{version}", shell_output("#{bin}/crossplane version --client")

```

**Review by chenrui333 — APPROVED** [review]
need to squash the commits

**Comment by Moulick** [comment]
@chenrui333 is the open thread the reason for waiting to merge?

**Review by nandahkrishna — APPROVED** [review]

**Pull request merged**